# What's cooking competiton

**The competitions** asks to predict the category of a dish's cuisine given a list of its ingredients.




In the dataset, we include the **recipe id**, **the type of cuisine**, and **the list of ingredients of each recipe** (of variable length). The data is stored in JSON format. 

#### An example of a recipe node in train.json:

```json
 {
 "id": 24717,
 "cuisine": "indian",
 "ingredients": [
     "tumeric",
     "vegetable stock",
     "tomatoes",
     "garam masala",
     "naan",
     "red lentils",
     "red chili peppers",
     "onions",
     "spinach",
     "sweet potatoes"
 ]
 },
 ```
 
In the test file **test.json**, the format of a recipe is the same as **train.json**, only *the cuisine type* is removed, <font color=red>as it is the target variable you are going to predict.</font>

#### File descriptions:

**train.json -** the training set containing recipes id, type of cuisine, and list of ingredients

**test.json -** the test set containing recipes id, and list of ingredients

**sample_submission.csv -** a sample submission file in the correct format
In the dataset, we include the recipe id, the type of cuisine, and the list of ingredients of each recipe (of variable length). The data is stored in JSON format. 

-------

In [1]:
# include required libraries
import os
import sys
import data_utils as du
import numpy as np
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split

Using TensorFlow backend.


In [2]:
# Read the data from files
trainingData = du.readJson('./data/train.json')
testData = du.readJson('./data/test.json')

### Exploring first few rows of each data set

In [3]:
print('TRAINING DATA:')
trainingData.head()

TRAINING DATA:


cuisine     id                                        ingredients
0        greek  10259  [romaine lettuce, black olives, grape tomatoes...
1  southern_us  25693  [plain flour, ground pepper, salt, tomatoes, g...
2     filipino  20130  [eggs, pepper, salt, mayonaise, cooking oil, g...
3       indian  22213                [water, vegetable oil, wheat, salt]
4       indian  13162  [black pepper, shallots, cornflour, cayenne pe...

In [4]:
print('TEST DATA:')
testData.head()

TEST DATA:


id                                        ingredients
0  18009  [baking powder, eggs, all-purpose flour, raisi...
1  28583  [sugar, egg yolks, corn starch, cream of tarta...
2  41580  [sausage links, fennel bulb, fronds, olive oil...
3  29752  [meat cuts, file powder, smoked sausage, okra,...
4  35687  [ground black pepper, salt, sausage casings, l...

### Make a Distribution map for ingredients and food types in the training set

In [5]:
ingredientsColumn = trainingData['ingredients']
recipeType = trainingData['cuisine']

ingredientsDist = du.dataDistributionMap(ingredientsColumn)
cuisineDist = du.dataDistributionMap(recipeType)

print ('Number of cuisines in the data set:',len(cuisineDist))
print ('Number of ingredients in the data set:',len(ingredientsDist))

<class 'list'>
<class 'str'>
Number of cuisines in the data set: 20
Number of ingredients in the data set: 6714


-------

### Prepare and the data to be used as input of a Neural Network ase well as create training and dev data-set

1. we need to create input vectors representing each **recipe** and output vectors representing a **cuisine type**.
2. XYZ 
3. XYZ
4. XYZ

### Prepre the input matrix: 

In [6]:
# create a map of ingredients
w2i, i2w = du.wordsToMap(list(ingredientsDist.keys()))

In [7]:
max(len(recipe) for recipe in trainingData['ingredients'])


65

In [8]:
# An input matrix with having the ingredients encoded for each recipe and zero padding.
# Each row is a recipe in the training set and will be the input of the first layer of NN
X = du.convertToInputMatrix(trainingData['ingredients'], w2i)

### Sanity Check:

In [9]:
X[3]

array([0, 0, 0, ..., 0, 0, 0])

In [10]:
trainingData['ingredients'][0]

['romaine lettuce',
 'black olives',
 'grape tomatoes',
 'garlic',
 'pepper',
 'purple onion',
 'seasoning',
 'garbanzo beans',
 'feta cheese crumbles']

In [11]:
print(i2w[0],"\n",i2w[1], "\n",i2w[2], "\n",i2w[3])

thai basil 
 rice paddy herb 
 crimini mushrooms 
 part-skim mozzarella


### Prepare the expected output matrix:

In [12]:
# create a map of index to cuisine  nad cuisine to index
c2i, i2c = du.wordsToMap(list(cuisineDist))

In [13]:
Y = du.convertToOutputMatrix(trainingData['cuisine'], c2i)

### Sanity Check:


In [14]:
Y[0:20]

array([[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
       [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1, 0, 0,

In [15]:
trainingData['cuisine'][1] 

'southern_us'

In [16]:
Y[1]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0])

In [17]:
print(c2i['southern_us'], i2c[1])

15 brazilian


In [18]:
# split into 85% for train and 15% for test
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.15)

### Build the model:

In [19]:
# create model
model = Sequential()
model.add(Dense(32, input_dim=6714, activation='relu', kernel_initializer="uniform"))
model.add(Dense(44, activation='tanh', kernel_initializer="uniform"))
model.add(Dense(32, activation='tanh', kernel_initializer="uniform"))
model.add(Dense(32, activation='tanh', kernel_initializer="uniform"))
model.add(Dense(25, activation='tanh', kernel_initializer="uniform"))
model.add(Dense(20, activation='softmax'))

# Compile model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [20]:
# Fit the model
model.fit(X_train, Y_train, epochs=150, batch_size=500)

Epoch 1/150
33807/33807 [==============================] - 20s 603us/step - loss: 2.7013 - acc: 0.1942
Epoch 2/150
33807/33807 [==============================] - 2s 45us/step - loss: 2.2723 - acc: 0.2186
Epoch 3/150
33807/33807 [==============================] - 2s 54us/step - loss: 2.0206 - acc: 0.2938
Epoch 4/150
33807/33807 [==============================] - 2s 55us/step - loss: 1.8537 - acc: 0.3585
Epoch 5/150
33807/33807 [==============================] - 2s 56us/step - loss: 1.7121 - acc: 0.4183
Epoch 6/150
33807/33807 [==============================] - 2s 53us/step - loss: 1.5748 - acc: 0.4456
Epoch 7/150
33807/33807 [==============================] - 2s 55us/step - loss: 1.4605 - acc: 0.4928
Epoch 8/150
33807/33807 [==============================] - 2s 54us/step - loss: 1.3462 - acc: 0.5460
Epoch 9/150
33807/33807 [==============================] - 2s 54us/step - loss: 1.2208 - acc: 0.5956
Epoch 10/150
33807/33807 [==============================] - 2s 54us/step - loss: 1.1392 -

33807/33807 [==============================] - 2s 55us/step - loss: 0.2783 - acc: 0.9308
Epoch 82/150
33807/33807 [==============================] - 2s 50us/step - loss: 0.2747 - acc: 0.9335
Epoch 83/150
33807/33807 [==============================] - 2s 52us/step - loss: 0.2732 - acc: 0.9328
Epoch 84/150
33807/33807 [==============================] - 2s 54us/step - loss: 0.2667 - acc: 0.9360
Epoch 85/150
33807/33807 [==============================] - 2s 55us/step - loss: 0.2634 - acc: 0.9366
Epoch 86/150
33807/33807 [==============================] - 2s 55us/step - loss: 0.2606 - acc: 0.9369
Epoch 87/150
33807/33807 [==============================] - 2s 53us/step - loss: 0.2605 - acc: 0.9367
Epoch 88/150
33807/33807 [==============================] - 2s 55us/step - loss: 0.2573 - acc: 0.9372
Epoch 89/150
33807/33807 [==============================] - 2s 53us/step - loss: 0.2535 - acc: 0.9389
Epoch 90/150
33807/33807 [==============================] - 2s 52us/step - loss: 0.2515 - acc: 

In [21]:
# evaluate the model
scores = model.evaluate(X_test, Y_test)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

5967/5967 [==============================] - 0s 82us/step

acc: 53.46%


There is a bias between training and test 95.65% vs 53.46%